core/camel-base-engine/src/main/java/org/apache/camel/impl/engine/CamelInternalProcessor.java

```diff
@@ -731,6 +731,7 @@ public UnitOfWork before(Exchange exchange) throws Exception {
                     this.routeId = route.getRouteId();
                 }
                 ExtendedExchange ee = (ExtendedExchange) exchange;
+                ee.setFromEndpoint(route.getEndpoint());
```

Comment:
Whats the use case for this? 

The exchange could already have been assigned with an endpoint, so you would potentially override this.

Reply:
Thanks for your feedback. I was just writing the comment below at the same time as you commented on the change here.